source/extensions/filters/http/peer_metadata/filter.cc

```diff
@@ -0,0 +1,344 @@
+  }
+  return {};
+}
+
+absl::optional<PeerInfo> XDSMethod::derivePeerInfo(const StreamInfo::StreamInfo& info,
+                                                   Http::HeaderMap&) const {
+  if (!metadata_provider_) {
+    return {};
+  }
+  Network::Address::InstanceConstSharedPtr peer_address;
+  if (downstream_) {
+    peer_address = info.downstreamAddressProvider().remoteAddress();
+  } else {
+    if (info.upstreamInfo().has_value()) {
+      auto upstream_host = info.upstreamInfo().value().get().upstreamHost();
+      if (upstream_host) {
+        const auto address = upstream_host->address();
+        switch (address->type()) {
+        case Network::Address::Type::Ip:
+          peer_address = upstream_host->address();
+          break;
+        case Network::Address::Type::EnvoyInternal:
+          if (upstream_host->metadata()) {
+            const auto& filter_metadata = upstream_host->metadata()->filter_metadata();
+            const auto& it = filter_metadata.find("tunnel");
+            if (it != filter_metadata.end()) {
+              const auto& destination_it = it->second.fields().find("destination");
+              if (destination_it != it->second.fields().end()) {
+                peer_address = Network::Utility::parseInternetAddressAndPortNoThrow(
+                    destination_it->second.string_value(), /*v6only=*/false);
+              }
+            }
+          }
+          break;
+        default:
+          break;
+        }
+      }
+    }
+  }
+  const auto metadata_object = metadata_provider_->GetMetadata(peer_address);
+  if (metadata_object) {
+    return Istio::Common::convertWorkloadMetadataToFlatNode(metadata_object.value());
+  }
+  return {};
+}
+
+absl::optional<PeerInfo> MXMethod::derivePeerInfo(const StreamInfo::StreamInfo&,
+                                                  Http::HeaderMap& headers) const {
+  const auto peer_id_header = headers.get(Headers::get().ExchangeMetadataHeaderId);
+  absl::string_view peer_id =
+      peer_id_header.empty() ? "" : peer_id_header[0]->value().getStringView();
+  const auto peer_info_header = headers.get(Headers::get().ExchangeMetadataHeader);
+  absl::string_view peer_info =
+      peer_info_header.empty() ? "" : peer_info_header[0]->value().getStringView();
+  if (!peer_info.empty()) {
+    return lookup(peer_id, peer_info);
+  }
+  return {};
+}
+
+void MXMethod::remove(Http::HeaderMap& headers) const {
+  headers.remove(Headers::get().ExchangeMetadataHeaderId);
+  headers.remove(Headers::get().ExchangeMetadataHeader);
+}
+
+absl::optional<PeerInfo> MXMethod::lookup(absl::string_view id, absl::string_view value) const {
+  // This code is copied from:
+  // https://github.com/istio/proxy/blob/release-1.18/extensions/metadata_exchange/plugin.cc#L116
+  if (max_peer_cache_size_ > 0 && !id.empty()) {
+    auto it = cache_.find(id);
+    if (it != cache_.end()) {
+      return it->second;
+    }
+  }
+  const auto bytes = Base64::decodeWithoutPadding(value);
```

Comment:
Why "WithoutPadding" here?

Reply:
For backwards compatibility with "Wasm" implementation of the same logic [here](https://github.com/istio/proxy/blob/master/extensions/metadata_exchange/plugin.cc#L133). Without padding is more forgiving so we have to keep it lax like that.